### PR TITLE
Added Streaming functionality for Safari Browsers

### LINKED ISSUE #418: GSoC Infragram.org full-screen UI and video upload Discussion and Planning
Hi, everyone! This is the Planning Issue for the [Infragram.org: full-screen UI and video upload](https://summerofcode.withgoogle.com/programs/2022/projects/rP8hthBg) for [infragram.org](https://infragram.org/). The project is in collaboration with the [Outreachy proposal: Design a new full-screen user interface for Infragram.org](https://publiclab.org/notes/stefq1111/04-08-2022/outreachy-proposal) . All are open for discussion and improvement.

**Process Video Locally**
- [x] Allow conversion of entire video.
- [x] Implement video controls on canvas.
- [x] Include another Canvas Element ( second Canvas ).
- [x] Record  video on process at second canvas.
- [x]  Save processed video file.
- [x] Integrate the above functionalities to with UI of outreachy collaborator. @stephaniequintana 

**Allow selecting multiple output resolutions as well**
- [x] Dynamically change the size of second canvas to fit required resolution.
- [x] Integrate the above functionalities to with UI of outreachy collaborator.

**Integrate Teamsys latest camera selection API** 
 Issue #39
- [ ] Download Temasys/AdapterJS library.
- [ ] Change all instances of the word `webRtcOptions.context ` to `webRtcDetectedType`
- [ ] Remove code `if (webRtcOptions.context === "flash")`
[Reference 1](https://github.com/publiclab/spectral-workbench.js/pull/172)  [Reference 2]( https://github.com/publiclab/spectral-workbench.js/issues/87)

**Accept multiple webcam resolutions**
- [x] Maintain css canvas size.
- [x] Change HTML canvas size to fit requested resolution.
- [x] Integrate the above functionalities to with UI of outreachy collaborator.

**Write UI tests for new Infragram UI Jest** 
- [ ] write test for infragram UI functionalities.

**Allow zooming/panning of video**
- [x]  Zooming :  Add uScale of uniform vect2 to shader.frag and shader.vert. 
- [ ]  Panning :  Add uPan of uniform vect2 to shader.frag and shader.vert .
- [ ] Implement above functionalities to infragram drawscreen function.
- [ ] Integrate the above functionalities to with UI of outreachy collaborator.

**Miscellaneous**

- [ ]  Add option to close webcam input
- [ ]  Enable image and video stream mirroring
- [x]  Add file filters on input to optimize infrared analysis

### PR DESCRIPTION
**Pull request by Forchapeatl**
- Removed Webrtc Library
-Added `getUserMedia` Api
-Added function  to initialize  video streaming  `src/io/camera.js` on `function unInitialize()`
Fixes #418 

Hosted on [https://forchapeatl.github.io/infragram/indexs.html](https://forchapeatl.github.io/infragram/indexs.html)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,11 +25,9 @@
     "bootstrap": "~3.4.0",
     "bootstrap5": "npm:bootstrap@^5.1.3",
     "font-awesome": "~4.7.0",
-    "getusermedia-js": "~1.0.0",
```
Comment (jywarren) [inline package.json:28]: Ah, if we remove these here, won't it break v1? Shall we leave them in even though v2 won't be using them?

Comment (jywarren) [reply package.json:28]: Or alternatively, should we implement the same changes in index2.html into index.html and fix both v1 and v2 in this respect?

Comment (Forchapeatl) [reply package.json:28]: OH sorry

Comment (Forchapeatl) [reply package.json]: Let's not break v1 for now. We can break it after v2 is complete , when @stephaniequintana and I have a synchronized codebase.

Comment (Forchapeatl) [reply package.json]: @jywarren , package.json has been fixed. Please may I merge?

**Comment by jywarren** [comment]
Can you check if any of the changes to the src files adversely affected v1
functionality? Thank you!

**Comment by stephaniequintana** [comment]
FANTASTIC work, @Forchapeatl! Kudos!!! 🚀 🎆

**Comment by Forchapeatl** [comment]
sorry about that @jywarren it affects the functionality of `camera.js` a whole lot. As the bootstrap `modal` is replaced by `offcanvas` and `WebcamVideo` element is not created by ` WebRTC`. Should I  create a seperate `cameraV2.js` file ?

**Comment by jywarren** [comment]
> sorry about that @jywarren it affects the functionality of `camera.js` a whole lot. As the bootstrap `modal` is replaced by `offcanvas` and `WebcamVideo` element is not created by ` WebRTC`. Should I create a seperate `cameraV2.js` file ?

It's ok! @stephaniequintana would you like to explain how we could use the options flag to switch behaviors between the two versions?

File: src/io/camera.js [hunk 2]
```diff
@@ -3,14 +3,12 @@
 module.exports = function Camera(options) {
   var canvas, 
       ctx;
-
+  options.webcamVideoSelector = options.webcamVideoSelector || 'webcamVideoEl';
+  options.webcamVideoEl = document.getElementById(options.webcamVideoSelector);
+  
   // Initialize getUserMedia with options
   function initialize() {
-    getUserMedia(webRtcOptions, success, deviceError);
```
Comment (jywarren) [inline src/io/camera.js:9]: For example here!

Comment (Forchapeatl) [reply src/io/camera.js:9]:
```diff
  function initialize() {
+     if (options.version ==1){
    getUserMedia(webRtcOptions, success, deviceError);
+   }else{
}
```

Comment (Forchapeatl) [reply src/io/camera.js:9]: I think i get it now .

Comment (stephaniequintana) [reply src/io/camera.js:9]: I believe what @Forchapeatl [just now updated](https://github.com/publiclab/infragram/pull/442/files#r939453098) should work:

```javascript
  function initialize() {
+     if (options.version == '1'){
    getUserMedia(webRtcOptions, success, deviceError);
+   }else{
}
```

Comment (jywarren) [reply src/io/camera.js:9]: Yes exactly!!

Comment (jywarren) [reply src/io/camera.js:9]: Would you mind testing it out, then let me know when you're ready!

Comment (jywarren) [reply src/io/camera.js:9]: And be careful, JavaScript has some odd behaviors when comparing different types. I think the flag is an integer, not a string, so double check!

**Comment by Forchapeatl** [comment]
Hello @jywarren the `options.version` condition has been included to the `camera.js`  and `dist/infragram.js` files. Please have a look at the video below . hosted on (version 2 )[https://forchapeatl.github.io/infragram/indexs.html](https://forchapeatl.github.io/infragram/indexs.html) and (version 1) [https://forchapeatl.github.io/infragram/index.html](https://forchapeatl.github.io/infragram/indexs.html)

https://user-images.githubusercontent.com/24577149/183268935-86ef963f-2f03-47b5-9887-bc6a909d0bc9.mp4

File: src/io/camera.js [hunk 3]
```diff
@@ -99,31 +123,46 @@ module.exports = function Camera(options) {
...
+
   // not doing anything now... for copying to a 2nd canvas
   function getSnapshot() {
-    var video;
-    // If the current context is WebRTC/getUserMedia (something
-    // passed back from the shim to avoid doing further feature
-    // detection), we handle getting video/images for our canvas 
-    // from our HTML5 <video> element.
-    if (webRtcOptions.context === "webrtc") {
+    if(options.version == 1){
```
Comment (jywarren) [inline src/io/camera.js:133]: This all looks great! Please do try to standardize spacing -- there are some auto-formatters you can use too to achieve this! So, `if (options.version == 1) {` <= note the spaces. But not a big deal -- it just is nice over the long term (over years) to have the code be very regularly formatted. 

Thanks @Forchapeatl!!!

File: dist/infragram2.js [hunk 4]
```diff
@@ -394,15 +396,20 @@ function _slicedToArray(arr, i) { if (Array.isArray(arr)) { return arr; } else i
     // This file was adapted from infragram-js:
     // http://github.com/p-v-o-s/infragram-js.
     module.exports = function Camera(options) {
-      var canvas, ctx; // Initialize getUserMedia with options
+      var canvas, ctx;
+      options.webcamVideoSelector = options.webcamVideoSelector || 'webcamVideoEl';
+      options.webcamVideoEl = document.getElementById(options.webcamVideoSelector); // Initialize getUserMedia with options
```
Comment (jywarren) [inline dist/infragram2.js:401]: very nice

File: src/io/camera.js (hunk 3 above)
Comment (Forchapeatl) [reply src/io/camera.js:133]: Ohh, please do you mean indenting ?

**Comment by jywarren** [comment]
So, one more thing @Forchapeatl -- here, are you generating the `dist/infragram.js` and then renaming it `dist/infragram2.js`? If all the version differences are managed by the `options.version` conditionals now, can't we begin leaving it as `dist/infragram.js` instead of infragram2.js, since both versions will be able to run off the same file? 

And thank you for merging this and moving it along -- i think sorry it might have needed just one more review before doing so but it's not a big deal since you did confirm that the previous version was working fine!

However if you open a new PR where we just use `dist/infragram.js` we'll have to be extra sure all features work in both versions since they'll all be running off the same file from there onward. Make sense? 

Thanks!!!

**Comment by jywarren** [comment]
Also just cc'ing @stephaniequintana can you and @Forchapeatl just put a 👍 on the above so I know we're all coordinated? Thank you! Exciting!!!!

File: src/io/camera.js (hunk 3 above)
Comment (jywarren) [reply src/io/camera.js:133]: Ah no, actually just notice there's a space between `if` and `(` in line 109, but on 133 that's missing. it's a very small thing and not super important but is kind of nice to preserve completely consistent formatting as much as possible. Thank you!

**Comment by jywarren** [comment]
And thanks for your patience this week as I'm teaching long hours so less able to reply quickly! Definitely continue relying on reviews from each other and other mentors as well!

**Pull request merged**